<?xml version="1.0"?>
<launch>
  <arg name="robot" default="default_robot"/>
  <arg name="ns" default="$(arg robot)"/>
  <arg name="target" default="localhost"/>
  <arg name="model" default="$(find s4_description)/xacro/s4_robot.xacro"/>
  <param name="robot_description" command="$(find xacro)/xacro --inorder $(arg model)" />

  <group ns="$(arg robot)">
    <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher" >
      <param name="tf_prefix" value="$(arg robot)"/>
    </node>

    <group ns="omni0">
      <group ns="wheel0">
        <node name="s4_omni_wheel_phycon" pkg="s4_omni" type="s4_omni_wheel_phycon" output="screen">
          <remap from="CANLink_out" to="/$(arg ns)/comport0/CANLink_out"/>
          <remap from="CANLink_in"  to="/$(arg ns)/comport0/CANLink_in"/>
          <param name="CAN_CH" value="A"/>
          <param name="CAN_ID" value="1"/>
          <param name="PPR" value="-397"/>
        </node>
      </group>
      <group ns="wheel1">
        <node name="s4_omni_wheel_phycon" pkg="s4_omni" type="s4_omni_wheel_phycon" >
          <remap from="CANLink_out" to="/$(arg ns)/comport0/CANLink_out"/>
          <remap from="CANLink_in"  to="/$(arg ns)/comport0/CANLink_in"/>
          <param name="CAN_CH" value="A"/>
          <param name="CAN_ID" value="2"/>
          <param name="PPR" value="-397"/>
        </node>
      </group>
      <group ns="wheel2">
        <node name="s4_omni_wheel_phycon" pkg="s4_omni" type="s4_omni_wheel_phycon" >
          <remap from="CANLink_out" to="/$(arg ns)/comport0/CANLink_out"/>
          <remap from="CANLink_in"  to="/$(arg ns)/comport0/CANLink_in"/>
          <param name="CAN_CH" value="A"/>
          <param name="CAN_ID" value="3"/>
          <param name="PPR" value="-397"/>
        </node>
      </group>
    </group>


    <node name="fixed_joint_publisher" pkg="s4_description" type="s4_description_joint_publisher" />

    <group ns="gun0">
      <node name="s4_gun_joint_phycon" pkg="s4_gun" type="s4_gun_joint_phycon" >
        <remap from="CANLink_out" to="/$(arg ns)/comport0/CANLink_out"/>
        <param name="CAN_CH" value="A"/>
        <param name="CAN_ID" value="0"/>
        <!--
        <param name="y_ajust" value="0.055"/>
        <param name="z_ajust" value="-0.045"/>
        -->
      </node>
      <node name="s4_gun_laser_phycon" pkg="s4_gun" type="s4_gun_laser_phycon" >
        <remap from="CANLink_out" to="/$(arg ns)/comport0/CANLink_out"/>
        <param name="CAN_CH" value="A"/>
        <param name="CAN_ID" value="0"/>
      </node>
      <node name="s4_gun_shot_phycon" pkg="s4_gun" type="s4_gun_shot_phycon" >
        <remap from="CANLink_out" to="/$(arg ns)/comport0/CANLink_out"/>
        <remap from="CANLink_in"  to="/$(arg ns)/comport0/CANLink_in"/>
        <param name="CAN_CH" value="A"/>
        <param name="CAN_ID" value="0"/>
      </node>
    </group>

    <group ns="sensor0">
      <group ns="front_lidar">
        <node machine="$(arg target)" name="ydlidar_node"  pkg="ydlidar"  type="ydlidar_node" output="screen">
          <param name="port"         type="string" value="/dev/ydlidar"/>
          <param name="baudrate"     type="int"    value="115200"/>
          <param name="frame_id"     type="string" value="$(arg robot)/sensor0/front_lidar_link"/>
          <param name="angle_fixed"  type="bool"   value="true"/>
          <param name="low_exposure"  type="bool"   value="false"/>
          <param name="heartbeat"    type="bool"   value="false"/>
          <param name="resolution_fixed"    type="bool"   value="true"/>
          <param name="angle_min"    type="double" value="-180" />
          <param name="angle_max"    type="double" value="180" />
          <param name="range_min"    type="double" value="0.08" />
          <param name="range_max"    type="double" value="16.0" />
          <param name="ignore_array" type="string" value="" />
          <param name="samp_rate"    type="int"    value="9"/>
          <param name="frequency"    type="double" value="7"/>

          <param name="reversion"    type="bool"   value="false"/>
<!--
          <param name="auto_reconnect"    type="bool"   value="true"/>
-->
        </node>
      </group>
      <group ns="head_camera">
        <node machine="$(arg target)" name="uvc_camera_node"  pkg="uvc_camera"  type="uvc_camera_node" output="screen">
          <param name="width" value="800"/>
          <param name="height" value="600"/>
          <param name="fps" value="10"/>
        </node>
      </group>
    </group>
  </group>

  <include file="$(find s4_comport)/launch/main.launch" >
    <arg name="target" value="$(arg target)"/>
    <arg name="ns"     value="$(arg robot)"/>
    <arg name="unit"   value="comport0"/>
    <arg name="device_name" value="/dev/CANadapter"/>
  </include>
</launch>

